packages/nouns-contracts/contracts/NounsDescriptorV2.sol

```diff
@@ -0,0 +1,427 @@
+     */
+    function toggleDataURIEnabled() external override onlyOwner {
+        bool enabled = !isDataURIEnabled;
+
+        isDataURIEnabled = enabled;
+        emit DataURIToggled(enabled);
+    }
+
+    /**
+     * @notice Set the base URI for all token IDs. It is automatically
+     * added as a prefix to the value returned in {tokenURI}, or to the
+     * token ID if {tokenURI} is empty.
+     * @dev This can only be called by the owner.
+     */
+    function setBaseURI(string calldata _baseURI) external override onlyOwner {
+        baseURI = _baseURI;
+
+        emit BaseURIUpdated(_baseURI);
+    }
+
+    /**
+     * @notice Given a token ID and seed, construct a token URI for an official Nouns DAO noun.
+     * @dev The returned value may be a base64 encoded data URI or an API URL.
+     */
+    function tokenURI(uint256 tokenId, INounsSeeder.Seed memory seed) external view override returns (string memory) {
+        if (isDataURIEnabled) {
+            return dataURI(tokenId, seed);
+        }
+        return string(abi.encodePacked(baseURI, tokenId.toString()));
+    }
+
+    /**
+     * @notice Given a token ID and seed, construct a base64 encoded data URI for an official Nouns DAO noun.
+     */
+    function dataURI(uint256 tokenId, INounsSeeder.Seed memory seed) public view override returns (string memory) {
+        string memory nounId = tokenId.toString();
+        string memory name = string(abi.encodePacked('Noun ', nounId));
+        string memory description = string(abi.encodePacked('Noun ', nounId, ' is a member of the Nouns DAO'));
+
+        return genericDataURI(name, description, seed);
+    }
+
+    /**
+     * @notice Given a name, description, and seed, construct a base64 encoded data URI.
+     */
+    function genericDataURI(
+        string memory name,
+        string memory description,
+        INounsSeeder.Seed memory seed
+    ) public view override returns (string memory) {
+        NFTDescriptorV2.TokenURIParams memory params = NFTDescriptorV2.TokenURIParams({
+            name: name,
+            description: description,
+            parts: getPartsForSeed(seed),
+            background: art.backgrounds(seed.background)
+        });
+        return NFTDescriptorV2.constructTokenURI(renderer, params);
+    }
+
+    /**
+     * @notice Given a seed, construct a base64 encoded SVG image.
+     */
+    function generateSVGImage(INounsSeeder.Seed memory seed) external view override returns (string memory) {
+        ISVGRenderer.SVGParams memory params = ISVGRenderer.SVGParams({
+            parts: getPartsForSeed(seed),
+            background: art.backgrounds(seed.background)
+        });
+        return NFTDescriptorV2.generateSVGImage(renderer, params);
+    }
+
+    /**
+     * @notice Get all Noun parts for the passed `seed`.
+     */
+    function getPartsForSeed(INounsSeeder.Seed memory seed) public view returns (ISVGRenderer.Part[] memory) {
+        bytes memory _body = art.bodies(seed.body);
```

Comment:
nit: rename to `body`
or: inline it into the struct construction, not really used otherwise

Reply:
removing the `_`. it's actually used later in `_getPalette`, so I can't inline. makes sense?